File: cursorless-talon/src/cheatsheet_html/sections/actions.py
from ...actions.actions import ACTION_LIST_NAMES
from ..get_list import get_raw_list, make_dict_readable


def get_actions():
    all_actions = {}
    for name in ACTION_LIST_NAMES:
        all_actions.update(get_raw_list(name))

    multiple_target_action_names = [
        "replaceWithTarget",
        "moveToTarget",
        "swapTargets",
        "applyFormatter",
        "wrapWithPairedDelimiter",
        "rewrap",
    ]
    simple_actions = {
        f"{key} <T>": value
        for key, value in all_actions.items()
        if value not in multiple_target_action_names
    }
    complex_actions = {
        value: key
        for key, value in all_actions.items()
        if value in multiple_target_action_names
    }

    swap_connective = list(get_raw_list("swap_connective").keys())[0]
    source_destination_connective = list(
        get_raw_list("source_destination_connective").keys()
    )[0]

    return [
        *make_dict_readable(
            "action",
            simple_actions,
            {
                "callAsFunction": "Call T on S",
            },
        ),
        {
            "identifier": "replaceWithTarget",
            "type": "action",
            "spokenForms": [
                {
                    "spokenForm": f"{complex_actions['replaceWithTarget']} <T1> {source_destination_connective} <T2>",
                    "description": "Replace T2 with T1",
                },
                {
                    "spokenForm": f"{complex_actions['replaceWithTarget']} <T>",
                    "description": "Replace S with T",
                },
            ],
        },
        {
            "identifier": "moveToTarget",
            "type": "action",
            "spokenForms": [
                {
                    "spokenForm": f"{complex_actions['moveToTarget']} <T1> {source_destination_connective} <T2>",
                    "description": "Move T1 to T2",
                },
                {
                    "spokenForm": f"{complex_actions['moveToTarget']} <T>",
                    "description": "Move T to S",
                },
            ],
        },
        {
            "identifier": "swapTargets",
            "type": "action",
            "spokenForms": [
                {
                    "spokenForm": f"{complex_actions['swapTargets']} <T1> {swap_connective} <T2>",
                    "description": "Swap T1 with T2",
                },
                {
                    "spokenForm": f"{complex_actions['swapTargets']} {swap_connective} <T>",
                    "description": "Swap S with T",
                },
            ],
        },
        {
            "identifier": "applyFormatter",
            "type": "action",
            "spokenForms": [
                {
                    "spokenForm": f"{complex_actions['applyFormatter']} <F> at <T>",
                    "description": "Reformat T as F",
                }
            ],
        },
        {
            "identifier": "wrapWithPairedDelimiter",
            "type": "action",
            "spokenForms": [
                {
                    "spokenForm": f"<P> {complex_actions['wrapWithPairedDelimiter']} <T>",
                    "description": "Wrap T with P",
                }
            ],
        },
        {
            "identifier": "rewrap",
            "type": "action",
            "spokenForms": [
                {
                    "spokenForm": f"<P> {complex_actions['rewrap']} <T>",
                    "description": "Rewrap T with P",
                }
            ],
        },
    ]

File: packages/common/src/spokenForms/actions.ts
import { ActionType } from "..";

const actions: Record<ActionType, string | null> = {
  scrollToBottom: "bottom",
  toggleLineBreakpoint: "break point",
  cutToClipboard: "carve",
  scrollToCenter: "center",
  clearAndSetSelection: "change",
  remove: "chuck",
  insertCopyBefore: "clone up",
  insertCopyAfter: "clone",
  toggleLineComment: "comment",
  copyToClipboard: "copy",
  scrollToTop: "crown",
  outdentLine: "dedent",
  revealDefinition: "define",
  editNewLineBefore: "drink",
  insertEmptyLineBefore: "drop",
  extractVariable: "extract",
  insertEmptyLineAfter: "float",
  foldRegion: "fold",
  followLink: "follow",
  deselect: "give",
  highlight: "highlight",
  showHover: "hover",
  indentLine: "indent",
  showDebugHover: "inspect",
  setSelectionAfter: "post",
  editNewLineAfter: "pour",
  setSelectionBefore: "pre",
  insertEmptyLinesAround: "puff",
  showQuickFix: "quick fix",
  showReferences: "reference",
  rename: "rename",
  reverseTargets: "reverse",
  findInWorkspace: "scout all",
  randomizeTargets: "shuffle",
  generateSnippet: "snippet make",
  sortTargets: "sort",
  setSelection: "take",
  revealTypeDefinition: "type deaf",
  unfoldRegion: "unfold",
  callAsFunction: "call",
  swapTargets: "swap",
  replaceWithTarget: "bring",
  moveToTarget: "move",
  wrapWithPairedDelimiter: "wrap",
  wrapWithSnippet: "wrap",
  rewrapWithPairedDelimiter: "repack",
  insertSnippet: "snippet",
  pasteFromClipboard: "paste",

  ["experimental.setInstanceReference"]: "from",

  editNew: null,
  executeCommand: null,
  getText: null,
  replace: null,

  // applyFormatter: "format",
  // findInDocument: "scout",
  // nextHomophone: "phones",
};

export function actionToSpokenForm(name: ActionType): string {
  if (name === "wrapWithSnippet") {
    name = "wrapWithPairedDelimiter";
  }
  const result = actions[name];
  if (result == null) {
    throw Error(`Unknown action '${name}'`);
  }
  return result;
}
